Renderer abstraction + SDL3 port

MontyTRC89 opened the pull request:
This pull request covers two areas:
- Renderer abstraction. Now renderer is totally API-unaware and It uses interfaces when it needs to deal with graphics API. I have then created a concrete DirectX 11 implementation, and in the future we'll have more.
- Move to SDL3. I have replaced most of Windows native code with SDL3 code, that is cross-platform. The native code now is mostly contained in the Windows subsystem class. In the future we'll have more subsystems (Linux, MacOS...).

We need to test that the renderer is working as expected and that all graphics effects work.
Additionally, we need to test the game in general, verifying if there're problems with multithreading, game initialization and finalization, configuration.

MontyTRC89 committed:
First interfaces

MontyTRC89 committed:
First classes of DX11 native implementation

MontyTRC89 committed:
Added more classes to DX11 native implementation

MontyTRC89 committed:
Almost completed interfaces abstraction

MontyTRC89 committed:
WIP

MontyTRC89 committed:
Merge branch 'develop' into renderer_abstraction

MontyTRC89 committed:
WIP 2

MontyTRC89 committed:
WIP 2

MontyTRC89 committed:
WIP 3

MontyTRC89 committed:
WIP 4

MontyTRC89 committed:
First partially working build

MontyTRC89 committed:
Fixed bugs, now scene draws with minor bugs

MontyTRC89 committed:
Fixed inventory not changing object on the left/right

MontyTRC89 committed:
Removed DX9 and DX11 dependencies from the renderer;

MontyTRC89 committed:
Merge branch 'develop' into renderer_abstraction

# Conflicts:
#	TombEngine/Renderer/RendererDraw.cpp
#	TombEngine/Renderer/RendererSprites.cpp

MontyTRC89 committed:
WIP move to unique_ptr and refactoring

MontyTRC89 committed:
Converted raw pointers to unique_ptr and handled delete of resources on level change

MontyTRC89 committed:
Merge branch 'develop' into renderer_abstraction

MontyTRC89 committed:
DX11 classes cleaning

MontyTRC89 committed:
Made shaders std140 ready

MontyTRC89 committed:
Started implementing subsystem abstraction

MontyTRC89 committed:
Added SDL3;
Created subsystems infrastructure;

MontyTRC89 committed:
Deleted more Windows specific code from main;
Using SDL thread library instead of Win32 API;

MontyTRC89 committed:
Fixed threading handling;
Cleaned more Windows specific code;

MontyTRC89 committed:
Copy SDL3.dll to target build directory;
Rewritten clock system with SDL3;
Removed more Windows specific code and replaced it with SDL3;
Renamed WinMain.h to engine_main.h;

MontyTRC89 committed:
Moved GPU swtch code to Windows subsystem;
Moved FreeConsole to Windows subsystem;

MontyTRC89 committed:
Read and write configuration from disk and not from registry; Added function for listing sound devices;

MontyTRC89 committed:
Removed setup dialog; Allow in-game sound device switch; Use SDL3 for configuration instead of Windows registry;

MontyTRC89 committed:
Merge branch 'develop' into renderer_abstraction

MontyTRC89 committed:
Set default character to whitespace in SpriteFont class;
Fixed init of default configuration missing;
Improved sound device reset in game;

MontyTRC89 committed:
Removed old D3D9 include

MontyTRC89 committed:
Merge branch 'develop' into renderer_abstraction;

MontyTRC89 committed:
Improved handling of sound disabled flag and reset also video and mixdown channels

MontyTRC89 committed:
Merge branch 'develop' into renderer_abstraction

MontyTRC89 committed:
Refactored width, height, windowed variables in renderer;
Refactored initialization;

MontyTRC89 committed:
Fixed video player;
Added SetScissor that accepts also a RendererVIewport

MontyTRC89 committed:
Removed unused variables for binding vertex buffers

MontyTRC89 committed:
Fixed windowed switch;
FIxed input bindings not saving to configuration;

MontyTRC89 committed:
Fixed texture content not initialized in DX11Texture2D

Sezzary commented on TombEngine/Game/control/control.h:
```diff
@@ -1,4 +1,5 @@
 #pragma once
+#include <sdl3/SDL.h>
```
Small first note, I encourage everyone to keep library includes like this in `framework.h` with the others. :D

Lwmte replied:
I think it's already included, so this one may be safely removed I guess?

MontyTRC89 replied:
Done. I will remove others when found. SDL3 is included in framework.h

MontyTRC89 committed:
Removed not needed SDL3 lib files

MontyTRC89 committed:
Merge branch 'develop' into renderer_abstraction

MontyTRC89 committed:
Merge remote-tracking branch 'origin/renderer_abstraction' into renderer_abstraction

MontyTRC89 committed:
engine_main renamed to EngineMain; Removed SDL3 included everywhere because lib is already included in frameowrk.h

MontyTRC89 commented on TombEngine/Sound/sound.cpp:
```diff
@@ -1086,6 +1094,67 @@ void Sound_DeInit()
+
+		// Clear the old device
+		BASS_SetDevice(oldSoundDevice);
+		BASS_Free();
```
SDL3 include fixed

MontyTRC89 replied:
I have added a new call to Bass_SetDevice at the end :-)

MontyTRC89 committed:
Merge branch 'develop' into renderer_abstraction

MontyTRC89 committed:
Restored CRS exe call;
Moved generate dummy level to platform specific code;

MontyTRC89 committed:
Fixed race condition that made the game crash due to spdlog terminated too early

MontyTRC89 committed:
Merge branch 'develop' into renderer_abstraction

MontyTRC89 committed:
Set again the new sound device at the end of Sound_Reset

Lwmte commented on TombEngine/Renderer/Native/DirectX11/DX11GraphicsDevice.cpp:
```diff
@@ -0,0 +1,1017 @@
+#include "framework.h"
+
+#ifdef SDL_PLATFORM_WIN32
```
Why this particular file is completely under this `ifdef`, but other DirectX-specific ones are not?

Lwmte commented on TombEngine/Renderer/Native/DirectX11/DX11SpriteFont.h:
```diff
@@ -0,0 +1,99 @@
+#pragma once
```
By the way, are we adopting OpenLara way of doing things, and prefer to keep all code in .h files now?

Lwmte commented on TombEngine/Renderer/RendererEnums.h:
```diff
@@ -1,17 +1,16 @@
-#define PRINTSTRING_COLOR_WHITE D3DCOLOR_ARGB(255, 255, 255, 255)
-#define PRINTSTRING_COLOR_BLACK D3DCOLOR_ARGB(255, 0, 0, 0)
-#define PRINTSTRING_COLOR_YELLOW D3DCOLOR_ARGB(255, 240, 220, 32)
+#define PRINTSTRING_COLOR_ORANGE 0xFFD87531u
```
Do not introduce magic numbers, please.

Lwmte commented on TombEngine/Renderer/RendererEnums.h:
```diff
@@ -83,6 +82,8 @@ constexpr auto SHININESS_VERTEX_SHIFT = 16;
 constexpr auto LOCKED_VERTEX_SHIFT = 24;
 constexpr auto INDEX_IN_POLY_VERTEX_SHIFT = 25;
 
+#define SAFE_DELETE(x) if (x != nullptr) x.reset();
```
This macro does not belong to Renderer**Enums**.h. Move it to appropriate place.

MontyTRC89 replied:
Done

Lwmte commented on TombEngine/Renderer/RendererEnums.h:
```diff
@@ -120,7 +121,8 @@ enum class CullMode
 	Clockwise = 1,
-	CounterClockwise = 2
+	CounterClockwise = 2,
+	Wireframe = 3
```
Why wireframe is a cull mode? I thought it is a rendering mode.

Lwmte commented on TombEngine/Renderer/RendererCompatibility.cpp:
```diff
@@ -61,44 +57,47 @@ namespace TEN::Renderer
+				std::move(emissive));
 
-			_animatedTextures[i] = tex;
+			_animatedTextures[i] = std::move(tex);
```
Why `std::move` is used here?

Lwmte commented on TombEngine/Renderer/RendererCompatibility.cpp:
```diff
@@ -147,45 +146,48 @@ namespace TEN::Renderer
+				std::move(emissive));
 			
-			_roomTextures[i] = tex;
+			_roomTextures[i] = std::move(tex);
```
As above.

Lwmte commented on TombEngine/Renderer/RendererDraw.cpp:
```diff
@@ -151,43 +148,66 @@ namespace TEN::Renderer
 			_shadowLight->Position :
 			Vector3::Lerp(_shadowLight->PrevPosition, _shadowLight->Position, GetInterpolationFactor());
 
+		static constexpr Vector3 forwardVectors[6] =
```
Can we use `Vector3::UnitX` etc here? Available DirectXMath constants are:

```            static const Vector3 Zero;
            static const Vector3 One;
            static const Vector3 UnitX;
            static const Vector3 UnitY;
            static const Vector3 UnitZ;
            static const Vector3 Up;
            static const Vector3 Down;
            static const Vector3 Right;
            static const Vector3 Left;
            static const Vector3 Forward;
            static const Vector3 Backward;

MontyTRC89 replied:
Done

Lwmte commented on TombEngine/Renderer/RendererDrawMenu.cpp:
```diff
@@ -129,6 +128,9 @@ namespace TEN::Renderer
 		sprintf(stringBuffer, "%d x %d", screenResolution.x, screenResolution.y);
 
+		char soundDeviceStringBuffer[255] = {};
+		sprintf(soundDeviceStringBuffer, "%s", g_Configuration.SupportedSoundDevices[g_Gui.GetCurrentSettings().SelectedSoundDevice].Name.c_str());
```
Sound device name may contain non-Latin characters (e.g. default sound device on non-English systems). Maybe a good idea is to use Unicode here?

Lwmte commented on TombEngine/Resources.rc:
```diff
@@ -27,8 +17,8 @@ LANGUAGE LANG_ENGLISH, SUBLANG_NEUTRAL
 VS_VERSION_INFO VERSIONINFO
- FILEVERSION TEN_MAJOR_VERSION,TEN_MINOR_VERSION,TEN_BUILD_NUMBER,TEN_REVISION_NUMBER
- PRODUCTVERSION TEN_MAJOR_VERSION, TEN_MINOR_VERSION, TEN_BUILD_NUMBER, TEN_REVISION_NUMBER
+ FILEVERSION 1,10,0,2
```
Is version now hardcoded, or this file is auto-recreated by some script? `TEN_MAJOR_VERSION` etc macros were automatically upgrading version number when `Version.h` was changed.

MontyTRC89 replied:
Already fixed

Lwmte commented on TombEngine/Sound/sound.h:
```diff
@@ -153,6 +153,14 @@ struct SoundSourceInfo
 	}
 };
 
+struct BassDevice {
```
Use formatting according to coding conventions, please (no `{` on the same line).

MontyTRC89 replied:
Already fixed

Lwmte commented on TombEngine/Specific/EngineMain.cpp:
```diff
@@ -0,0 +1,521 @@
+	g_Platform->CheckPrerequisites();
+	
+	// Initialize SDL3
+	if (!SDL_Init(SDL_INIT_VIDEO | SDL_INIT_AUDIO | SDL_INIT_EVENTS))
```
Why are we initializing audio and video SDL subsystems, if they aren't used? Or video subsystem is already used?

MontyTRC89 replied:
It's needed

Lwmte commented on TombEngine/Specific/EngineMain.cpp:
```diff
@@ -0,0 +1,521 @@
+	// Initialize SDL3
+	if (!SDL_Init(SDL_INIT_VIDEO | SDL_INIT_AUDIO | SDL_INIT_EVENTS))
+	{
+		// handle error
```
I guess we need to at least report an error here via non-TENLog means?

Lwmte commented on TombEngine/Specific/Video/Video.cpp:
```diff
@@ -377,8 +376,18 @@ namespace TEN::Video
+				RenderBackground();
+			}
+
+			/*if (_videoTexture && SUCCEEDED(_d3dContext->Map(_videoTexture, 0, D3D11_MAP_WRITE_DISCARD, 0, &mappedResource)))
```
Why this block is commented? Was the video texture tested after this was removed?

MontyTRC89 replied:
Video textures already reworked

Lwmte commented on TombEngine/Specific/clock.cpp:
```diff
@@ -15,32 +15,32 @@ HighFramerateSynchronizer g_Synchronizer;
-		_frameTime = (_currentTime.QuadPart - _lastTime.QuadPart) * 1000.0 / _frequency.QuadPart;
-		_lastTime = _currentTime;
+		_currentCounter = SDL_GetPerformanceCounter();
+		Uint64 diff = _currentCounter - _lastCounter;
```
Let's use data types according to coding conventions, please. Maybe `unsigned long long`?

Lwmte commented on TombEngine/Specific/clock.cpp:
```diff
@@ -80,44 +79,41 @@ void HighFramerateSynchronizer::Step()
-
-	LdSync = (double)fq.LowPart + ((double)fq.HighPart * (double)0xffffffff);
-	LdSync /= LdFreq;
+	Uint64 counter = SDL_GetPerformanceCounter();
```
Use types according to coding conventions, e.g. `unsigned long long`.

Lwmte commented on TombEngine/Specific/configuration.cpp:
```diff
@@ -18,290 +23,35 @@ using namespace TEN::Utils;
-	return true;
+	SDL_IOStream* rw = SDL_IOFromFile(path.c_str(), "rb");
+	if (!rw) return false;
+	Sint64 size = SDL_GetIOSize(rw);
```
Use types according to coding conventions, please.

MontyTRC89 replied:
Done

MontyTRC89 committed:
Merge branch 'renderer_abstraction' of github-personal:TombEngine/TombEngine into renderer_abstraction

MontyTRC89 commented on TombEngine/Renderer/Native/DirectX11/DX11GraphicsDevice.cpp:
```diff
@@ -0,0 +1,1017 @@
+	void DX11GraphicsDevice::BindRenderTarget(IRenderTarget2D* renderTarget, IDepthTarget* depthTarget)
+	{
+		auto dxRt = static_cast<DX11RenderTarget2D*>(renderTarget);
+		auto rtv = dxRt->GetD3D11RenderTargetView();
```
Done

MontyTRC89 committed:
Fixed variable names in graphics device class; Fixed wrong enum conversion in DX11InputLayout class; Added SDL32 WIN32 ifdef in all DX11 classes;

MontyTRC89 committed:
Moved macros in trutils.h; Restored ARGB_TO_UINT macro (ex D3DCOLOR).

MontyTRC89 commented on TombEngine/Renderer/Renderer.h:
```diff
@@ -306,61 +276,54 @@ namespace TEN::Renderer
 		// Special effects
 
-		//std::vector<Texture2D> _causticTextures;
+		//std::vector<ITexture2D*> _causticTextures;
```
Caustics are now in sprites sequences, already before my refactor

MontyTRC89 committed:
Deleted legacy caustics vector

MontyTRC89 committed:
Using Vector3 constants instead of manual initializations

MontyTRC89 commented on TombEngine/Renderer/RendererDraw.cpp:
```diff
@@ -151,43 +148,66 @@ namespace TEN::Renderer
 			auto shadowProjection = CCameraMatrixBuffer{};
 			shadowProjection.ViewProjection = view * projection;
-			UpdateConstantBuffer(shadowProjection, _cbCameraMatrices);
+			UpdateConstantBuffer(&shadowProjection, _cbCameraMatrices.get());
```
Because of std::unique_ptr

MontyTRC89 commented on TombEngine/Renderer/RendererDraw.cpp:
```diff
@@ -3343,25 +3328,25 @@ namespace TEN::Renderer
-		_context->ClearState();
-		_swapChain->Present(1, 0);
+		_graphicsDevice->ClearState();
+		_graphicsDevice->Present(); // 1 0
```
It's a memo for me, i will delete it after tests. It was the original arguments of Present() of DX11 context

MontyTRC89 committed:
Reverted previous fix; Added IsValid to ITexture2D and restored safety checks;

MontyTRC89 committed:
Formatting typos

MontyTRC89 commented on TombEngine/Renderer/RendererDrawMenu.cpp:
```diff
@@ -1279,66 +1280,48 @@ namespace TEN::Renderer
-			rect.bottom = rect.top+thumbWidth / aspectRatio;
+			rect.Left = _graphicsDevice->GetScreenWidth() - thumbWidth;
+			rect.Top = thumbY;
+			rect.Right = rect.Left+ thumbWidth;
```
Done

MontyTRC89 commented on TombEngine/Renderer/RendererInit.cpp:
```diff
@@ -94,161 +84,28 @@ namespace TEN::Renderer
-		//_tempRoomAmbientRenderTarget1 = RenderTarget2D(_device.Get(), ROOM_AMBIENT_MAP_SIZE, ROOM_AMBIENT_MAP_SIZE, DXGI_FORMAT_R8G8B8A8_UNORM, false, DXGI_FORMAT_D24_UNORM_S8_UINT);
-		//_tempRoomAmbientRenderTarget2 = RenderTarget2D(_device.Get(), ROOM_AMBIENT_MAP_SIZE, ROOM_AMBIENT_MAP_SIZE, DXGI_FORMAT_R8G8B8A8_UNORM, false, DXGI_FORMAT_D24_UNORM_S8_UINT);
-		//_tempRoomAmbientRenderTarget3 = RenderTarget2D(_device.Get(), ROOM_AMBIENT_MAP_SIZE, ROOM_AMBIENT_MAP_SIZE, DXGI_FORMAT_R8G8B8A8_UNORM, false, DXGI_FORMAT_D24_UNORM_S8_UINT);
-		//_tempRoomAmbientRenderTarget4 = RenderTarget2D(_device.Get(), ROOM_AMBIENT_MAP_SIZE, ROOM_AMBIENT_MAP_SIZE, DXGI_FORMAT_R8G8B8A8_UNORM, false, DXGI_FORMAT_D24_UNORM_S8_UINT);
```
Moved to DX11GraphicsDevice concrete class as discussed some weeks ago on chat. It was not needed to abstract the entire blend state architecture, we reference them by enum and let the native device handle them.

MontyTRC89 committed:
Fixed code in SDL IO according to code conventions

MontyTRC89 committed:
Used NO_VALUE in configuration

MontyTRC89 commented on TombEngine/Specific/configuration.cpp:
```diff
@@ -347,274 +97,154 @@ void InitDefaultConfiguration()
 
-				if (GetDWORDRegKey(inputKey, buffer, &tempKeyID, boundKeyID) != ERROR_SUCCESS)
+				int actionId = ToInt(key.substr(5), -1);
+				int keyId = ToInt(val, -1);
```
Done

MontyTRC89 committed:
Removed sample data, not needed anymore; Decondensed Trim function;

MontyTRC89 committed:
Merge remote-tracking branch 'origin/develop' into renderer_abstraction

Sezzary commented on TombEngine/Specific/EngineMain.h:
```diff
@@ -0,0 +1,26 @@
+void PauseGameThread();
+void ResumeGameThread();
+void WaitIfGamePaused();
+void EngineClose();
```
Remember newline at the end of the file.

Sezzary commented on TombEngine/Specific/trutils.cpp:
```diff
@@ -91,6 +91,40 @@ namespace TEN::Utils
+
+	bool ToBool(const std::string v, bool def)
+	{
+		if (v == "1" || v == "true" || v == "True" || v == "TRUE")  return true;
```
As above.

Sezzary commented on TombEngine/Specific/trutils.h:
```diff
@@ -17,6 +17,10 @@ namespace TEN::Utils
+	std::string Trim(std::string s);
+	bool StartsWith(const std::string s, const char* pref);
+	int ToInt(const std::string v, int def);
+	bool ToBool(const std::string v, bool def);
```
Avoid using single-letter parameter names.

Sezzary commented on TombEngine/Specific/Input/Input.cpp:
```diff
@@ -10,6 +10,7 @@
 #include "Sound/sound.h"
 #include "Specific/clock.h"
 #include "Specific/trutils.h"
+#include "Specific/EngineMain.h"
```
`EngineMain.h` include should be one line higher.

Sezzary commented on TombEngine/Sound/sound.h:
```diff
@@ -153,6 +153,14 @@ struct SoundSourceInfo
 	}
 };
 
+struct BassDevice {
```
`{` should be on the next line.

Sezzary commented on TombEngine/Sound/sound.h:
```diff
@@ -201,4 +210,5 @@ float Sound_DistanceToListener(Pose *position);
-bool  Sound_UpdateEffectAttributes(int index, float pitch, float gain);
\ No newline at end of file
+bool  Sound_UpdateEffectAttributes(int index, float pitch, float gain);
+std::vector<BassDevice> Sound_ListDevices();
```
Remember newline at file end.

Sezzary commented on TombEngine/Renderer/RendererFrame.cpp:
```diff
@@ -1,6 +1,5 @@
 #include "framework.h"
 #include "Renderer/Renderer.h"
- 
 #include "Game/animation.h"
```
A newline should come after the `Renderer.h` include.

MontyTRC89 committed:
Fixed variable names in EngineMain.cpp; Cleaned the quit process of the engine;

Sezzary committed:
Include corrections

Sezzary committed:
More touch-ups

MontyTRC89 committed:
Merge branch 'develop' into renderer_abstraction

Sezzary committed:
Format `configuration.cpp`

MontyTRC89 committed:
Merge branch 'develop' into renderer_abstraction

# Conflicts:
#	TombEngine/framework.h

MontyTRC89 committed:
Merge remote-tracking branch 'origin/develop' into renderer_abstraction

MontyTRC89 committed:
Merge branch 'develop' into renderer_abstraction

MontyTRC89 committed:
Fixed wrong merge

MontyTRC89 committed:
Added some comments

MontyTRC89 committed:
Merge branch 'develop' into renderer_abstraction

Sezzary committed:
Tidy up classes

MontyTRC89 committed:
Started moving code from h in cpp in DX11 native classes

MontyTRC89 committed:
Merge remote-tracking branch 'origin/renderer_abstraction' into renderer_abstraction

MontyTRC89 committed:
Moved code from header to cpp files in DX11 native implementation

MontyTRC89 committed:
Merge branch 'develop' into renderer_abstraction

MontyTRC89 committed:
Merge branch 'develop' into renderer_abstraction

MontyTRC89 committed:
Merge branch 'develop' into renderer_abstraction

MontyTRC89 committed:
Merge branch 'develop' into renderer_abstraction

# Conflicts:
#	TombEngine/Game/gui.cpp
#	TombEngine/Renderer/ConstantBuffers/ConstantBuffer.h
#	TombEngine/Renderer/Graphics/IndexBuffer.h
#	TombEngine/Renderer/Graphics/RenderTarget2D.h
#	TombEngine/Renderer/Graphics/RenderTargetCube.h
#	TombEngine/Renderer/Graphics/Texture2D.h
#	TombEngine/Renderer/Graphics/Texture2DArray.h
#	TombEngine/Renderer/Graphics/VertexBuffer.h
#	TombEngine/Renderer/Renderer.h
#	TombEngine/Renderer/RendererCompatibility.cpp
#	TombEngine/Renderer/RendererDrawMenu.cpp
#	TombEngine/Renderer/RendererSettings.cpp
#	TombEngine/Renderer/RendererUtils.cpp

MontyTRC89 committed:
VRAM tracker;
Moveed throwIfFailed in DX11 native area;
Added more debug info;

MontyTRC89 committed:
Use int instead of size_t

MontyTRC89 committed:
Restored external crash report system

MontyTRC89 committed:
Moved GetBinaryPath in subsystem;
Added Linux and MacOS subsystems;

MontyTRC89 committed:
Merge branch 'develop' into renderer_abstraction

MontyTRC89 committed:
Merge branch 'develop' into renderer_abstraction

MontyTRC89 committed:
Replaced ZeroMemory with memset; Replaced Windows message box API with SDL3; Replaced get assembly path with SDL3 version;

MontyTRC89 committed:
Made sound system cross-platform

MontyTRC89 committed:
Fixed CoInitializeEx() not called in Windows 7 and Windows 8

MontyTRC89 committed:
Removed references to HWND and DWORD where possible; Fixed crash on quit caused by VRAM tracked;

MontyTRC89 committed:
Merge branch 'develop' into renderer_abstraction

MontyTRC89 committed:
Call DX11 context flush method after allocating textures and render targets for avoiding crashes on some GPUs

MontyTRC89 committed:
Merge branch 'develop' into renderer_abstraction

MontyTRC89 committed:
Simplified Texture2D constructors

MontyTRC89 committed:
Simplified binding of constant buffers

MontyTRC89 committed:
Refactored bind shader functions

MontyTRC89 committed:
Merge branch 'develop' into renderer_abstraction

MontyTRC89 committed:
Cleaned check for 64 bit platform

MontyTRC89 committed:
Merge branch 'develop' into renderer_abstraction

MontyTRC89 committed:
Use PLATFORM_64BIT macro

MontyTRC89 committed:
Removed unused DX11 includes in not-platform code; Fixed some remaining Windows specific pieces of code;

MontyTRC89 committed:
Calculate MIP levels only for PNG textures

MontyTRC89 committed:
Merge branch 'develop' into renderer_abstraction

# Conflicts:
#	TombEngine/Renderer/Graphics/VRAMTracker.h

MontyTRC89 committed:
FIxed other 64 bit macros

MontyTRC89 committed:
Merge branch 'develop' into renderer_abstraction

MontyTRC89 committed:
Merge branch 'develop' into renderer_abstraction

MontyTRC89 committed:
Merge branch 'develop' into renderer_abstraction

MontyTRC89 committed:
Merge branch 'develop' into renderer_abstraction

MontyTRC89 committed:
Fix video playback with abstracted renderer and fix OtherSettings menu indices    Video:   - Add isDynamic parameter to CreateTexture2D for D3D11 DYNAMIC textures     required by Map/WRITE_DISCARD (video frame upload)   - Defer GPU texture creation from VLC thread to main thread to avoid     D3D11 context calls from non-main threads   - Update VLC args for VLC 4.0 compatibility: remove --vout=none (blocks     decoder device creation) and use --aout=amem for memory audio routing   - Fix infinite loop when VLC fails immediately (handle Stopping state)   - Fix memory leak: use unique_ptr::reset() instead of release()   - Remove leftover commented-out DX11 code    Menu:   - Fix OtherSettings menu highlight indices (Reverb through Cancel were     off by one, causing Sound Device and Reverb to flash together)

MontyTRC89 committed:
Merge branch 'develop' into renderer_abstraction

MontyTRC89 committed:
Fixed previous commit

Lwmte committed:
Merge branch 'develop' into renderer_abstraction

Lwmte committed:
Fix merge

MontyTRC89 committed:
Merge branch 'develop' into renderer_abstraction

MontyTRC89 committed:
Fixed previous commit

MontyTRC89 committed:
Fix sound device switching and missing sprite batch begin in string rendering                                                                                                                   Sound:                                                                                                                                                                                         - Re-apply BASS 3D factors (distance, rolloff, Doppler) on new device,     since these are per-device settings reset on new device init   - Move currently-playing sample channels to the new device before freeing     the old one, otherwise active channels get invalidated   - Call BASS_Apply3D() after switching to propagate 3D changes    Renderer:   - Add missing _spriteBatch->Begin() before drawing strings in RendererString

MontyTRC89 committed:
Fetch PE metadata version from version.h

FILEVERSION, PRODUCTVERSION and the FileVersion/ProductVersion strings
in Resources.rc were hardcoded to 1.10.0.2, causing the game to report
the wrong version through GetProductOrFileVersion() and triggering a
spurious "Level version is different from TEN version" warning.

Resources.rc now includes version.h and uses its macros, and the
duplicate (stale) version defines have been removed from resource.h
so that version.h is the single source of truth.

MontyTRC89 committed:
Fix sound device switching and restore menu alignment

Sound:
- Don't call BASS_Free when selecting "No sound device"; stop playback
  instead, keeping sample handles alive for re-enabling later.
- Recreate BASS_3D_Mixdown (STREAMPROC_DEVICE_3D) and BASS_Video on
  the new device instead of moving them, since they are device-bound.
- Re-apply BASS_CONFIG_3DALGORITHM on new device (per-device setting).
- Re-attach reverb and compressor FX after device switch.
- Track BASS init state with BASS_Initialized flag instead of relying
  on g_Configuration.EnableSound in Sound_DeInit.
- Extract Sound_InitDevice helper to share init logic between
  Sound_Init and Sound_Reset.

Menu:
- Restore MenuLeftSideEntry/MenuRightSideEntry to 180/520 matching
  develop branch.

MontyTRC89 committed:
Merge remote-tracking branch 'origin/renderer_abstraction' into renderer_abstraction

MontyTRC89 committed:
Fix music tracks not restarting after no-device round trip and remove duplicated BASS init code

- Pause soundtrack channels instead of stopping them when selecting
  "No sound device", so track info (Channel/Track) is preserved.
  When switching back to a real device, paused streams are moved
  and resumed automatically.
- Extract Sound_CreateMixdownStreams() and Sound_ApplyDeviceSettings()
  helpers, used by both Sound_Init and Sound_Reset, eliminating the
  duplicated mixdown/video/FX creation code.

MontyTRC89 committed:
Don't resume sounds when sound is disabled

ResumeAllSounds was blindly resuming paused soundtrack channels when
closing the pause menu, undoing the pause that Sound_Reset had just
applied for "No sound device". Early-out when EnableSound is false.

Lwmte committed:
Address some PR notes

Lwmte committed:
Use data types according to coding conventions

Lwmte committed:
Merge branch 'develop' into renderer_abstraction

Lwmte committed:
Delete unused LEB128 and chunk modules

Lwmte committed:
Rearrange and remove unnecessary SDL binaries

Lwmte committed:
Merge branch 'develop' into renderer_abstraction

MontyTRC89 committed:
Fix crash when re-entering level with video texture

_videoSprite.Texture was not reset in FreeRendererData(), leaving a
dangling pointer after level unload. On re-entry the renderer would
attempt to bind the stale texture and crash.

MontyTRC89 committed:
Merge remote-tracking branch 'origin/renderer_abstraction' into renderer_abstraction

MontyTRC89 committed:
Possibly fixed video playblack in silent mode

MontyTRC89 committed:
Fix string transparency fade curve and silent video playback

SpriteBatch:
- Add Begin() no-args overload matching DXTK default (premultiplied
  alpha blend, deferred sorting). SpriteFont text requires premultiplied
  blend; the previous explicit Begin(Deferred, AlphaBlend) was mapped
  to NonPremultiplied, causing alpha to be applied twice (squared curve).

RendererString:
- Call Begin() without arguments to match develop behavior exactly.

Video:
- Disable audio decoding via libvlc_media_add_option(":no-audio") when
  silent=true. The switch from --aout=adummy (develop) to --aout=amem
  (this branch) meant VLC falls back to system audio output when no
  memory callbacks are registered.

MontyTRC89 committed:
Guard VRAM tracker logs behind _DEBUG

Lwmte committed:
Use internal AlphaBlend blend mode instead of empty overload

MontyTRC89 committed:
Restored aout=adummy

MontyTRC89 committed:
Merge remote-tracking branch 'origin/renderer_abstraction' into renderer_abstraction

MontyTRC89 committed:
Fixed video's audio playing when sound disabled

MontyTRC89 committed:
Merge branch 'develop' into renderer_abstraction

MontyTRC89 committed:
Added missing call to load global vars in EngineMain.cpp

MontyTRC89 committed:
Merge branch 'develop' into renderer_abstraction

MontyTRC89 committed:
Fixed previous commit

Lwmte committed:
Use constants for option entry names

Lwmte commented on TombEngine/Specific/Video/Video.cpp:
```diff
@@ -587,6 +551,13 @@ namespace TEN::Video
 		player->_needRender = true;
 
+		static bool firstFrame = true;
+		if (firstFrame)
```
What is the purpose of this flag and debug entry? Is it a debug leftover?

MontyTRC89 replied:
Yes. removed.

Lwmte commented on TombEngine/Specific/Video/Video.cpp:
```diff
@@ -635,6 +606,8 @@ namespace TEN::Video
 
 	unsigned int VideoHandler::OnVideoSetup(void** data, char* chroma, unsigned* width, unsigned* height, unsigned* pitches, unsigned* lines)
 	{
+		TENLog("VLC OnVideoSetup: " + std::to_string(*width) + "x" + std::to_string(*height), LogLevel::Info);
```
Debug leftover?

MontyTRC89 replied:
yes. removed.

MontyTRC89 committed:
Remove leftover VLC debug traces

Lwmte committed:
Update Video.cpp

Lwmte committed:
Deslop

MontyTRC89 commented on TombEngine/Renderer/RendererSettings.cpp:
```diff
@@ -1,76 +1,48 @@
 		if (std::filesystem::is_regular_file(path))
 		{
-			texture = Texture2D(_device.Get(), path);
+			texture = _graphicsDevice->CreateTexture2D(TEN::Utils::ToString(path));
```
I have a branch ready for handling string-wstring thing. Could we merge now this branch and immediately after the new strings branch?

MontyTRC89 commented on TombEngine/Renderer/ShaderManager/ShaderManager.cpp:
```diff
@@ -174,8 +165,24 @@ namespace TEN::Renderer::Utils
+
 		// Helper function to load or compile a shader.
-		auto loadOrCompileShader = [this, type, defines, forceRecompile, shaderPath, compiledShaderPath]
+		/*auto loadOrCompileShader = [this, type, defines, forceRecompile, shaderPath, compiledShaderPath]
```
Deleted, it was there for reference

MontyTRC89 committed:
Removed commented code in shader manager

MontyTRC89 commented on TombEngine/Renderer/ShaderManager/ShaderManager.cpp:
```diff
@@ -295,35 +302,17 @@ namespace TEN::Renderer::Utils
 	{
 		Destroy(shader);
-		_shaders[(int)shader] = LoadOrCompile(fileName, funcName, type, defines, forceRecompile);
+		_shaders[(int)shader] = std::move(LoadOrCompile(fileName, funcName, type, defines, forceRecompile));
```
Overengineering. Removed.

MontyTRC89 committed:
Drop redundant std::move on LoadOrCompile return value

MontyTRC89 commented on TombEngine/Resources.rc:
```diff
@@ -48,8 +38,8 @@ BEGIN
             VALUE "ProductName", "Tomb Engine"
-            VALUE "FileVersion", TEN_VERSION_STRING
-            VALUE "ProductVersion", TEN_VERSION_STRING
+            VALUE "FileVersion", "1.10.0.2"
```
Already fixed

MontyTRC89 commented on TombEngine/Resources.rc:
```diff
@@ -66,70 +57,62 @@ END
 
 
+/////////////////////////////////////////////////////////////////////////////
+// Italiano (Italia) resources
```
Fixed for now

MontyTRC89 committed:
Strip Italian locale block from Resources.rc

The LANG_ITALIAN/SUBLANG_ITALIAN section was injected by the Visual
Studio Resource Editor when the .rc file was opened through its GUI
under an Italian system locale. The block is gated by APSTUDIO_INVOKED
so it is inert during normal rc.exe builds, but it keeps reappearing
in diffs whenever the file is saved through the Resource Editor.

Edit Resources.rc as plain text going forward to keep it from coming
back. Also normalise the remaining comment to English.

MontyTRC89 commented on TombEngine/Sound/sound.cpp:
```diff
@@ -1001,9 +1001,17 @@ void Sound_Init(const std::string& gameDirectory)
+	}
+
 	// HACK: Manually force-load ADPCM codec, because on Win11 systems it may suddenly unload otherwise.
 	ADPCMLibrary = LoadLibrary("msadp32.acm");
```
Backported the fix from develop_multiplatform in the past weeks

MontyTRC89 commented on TombEngine/Sound/sound.cpp:
```diff
@@ -1086,6 +1094,70 @@ void Sound_DeInit()
 
+void Sound_Reset()
+{
+	if (g_Configuration.EnableSound)
```
There were already many changes so it should not be valid this comment anymore

MontyTRC89 commented on TombEngine/Specific/EngineMain.cpp:
```diff
@@ -0,0 +1,521 @@
+	return arg == lowerName;
+}
+
+
```
Not valid anymore

MontyTRC89 commented on TombEngine/Specific/EngineMain.cpp:
```diff
@@ -0,0 +1,521 @@
+		}
+		else if (ArgEquals(arg.c_str(), "level") && i + 1 < argc)
+		{
+			levelFile = argv[++i]; // consumi l�argomento successivo
```
Already done

MontyTRC89 commented on TombEngine/Renderer/RendererPostProcess.cpp:
```diff
@@ -147,53 +142,36 @@ namespace TEN::Renderer
-		_context->RSSetScissorRects(1, rects);
+		RendererViewport viewport = { 0, 0, (int)( _graphicsDevice->GetScreenWidth() / factor), (int)( _graphicsDevice->GetScreenHeight() / factor), 0.0f, 1.0f };
+		_graphicsDevice->SetViewport(viewport);
+		_graphicsDevice->SetScissor(viewport);
```
I uniformed the calls and made everything symmetric.

MontyTRC89 commented on TombEngine/Renderer/RendererPostProcess.cpp:
```diff
@@ -203,120 +181,96 @@ namespace TEN::Renderer
-		_context->RSSetScissorRects(1, rects);
+		RendererViewport viewport = { 0, 0, (int)( _graphicsDevice->GetScreenWidth() / GLOW_DOWNSCALE_FACTOR), (int)( _graphicsDevice->GetScreenHeight() / GLOW_DOWNSCALE_FACTOR), 0.0f, 1.0f };
+		_graphicsDevice->SetViewport(viewport);
+		_graphicsDevice->SetScissor(viewport);
```
I uniformed the calls and made everything symmetric.

MontyTRC89 committed:
Log SDL_Init failure and initialize TENLog earlier

InitTENLog used to run after SDL_Init, so a failure at that step would
exit silently with code 1. Move arg parsing, GameDirectory resolution
and InitTENLog ahead of SDL_Init (none of them depend on SDL) so any
later failure is captured in the log file.

Also surface SDL_Init failures to the user via Platform::ShowErrorMessage
with the actual SDL_GetError() text, instead of the previous "Handle error"
placeholder comment.

Lwmte committed:
Merge branch 'develop' into renderer_abstraction

MontyTRC89 committed:
Shutdown TENLog on SDL_Init failure

MontyTRC89 committed:
Merge branch 'develop' into renderer_abstraction

MontyTRC89 committed:
Merge remote-tracking branch 'origin/renderer_abstraction' into renderer_abstraction

TrainWrack reviewed (commented):
I tried different resolutions and windowed/non windowed modes.
Smoke was visible for me.
I changed sound devices as well and they changed perfectly.
Played the whole level and did not notice anything different. 
Had on screen sprites and text.

Lwmte committed:
Revert pointer casts

Lwmte reviewed (approved):
Tested with shabaobab's Easter level for 2-3 hours, couldn't find any issues. Also eyeballed code 2-3 times, looks good
